test/e2e/runner/infra.go

```diff
@@ -0,0 +1,82 @@
+package main
+
+import (
+	"context"
+
+	e2e "github.com/tendermint/tendermint/test/e2e/pkg"
+)
+
+// TestnetInfraAPI provides an API for interacting with the infrastructure for
+// an entire testnet.
+type TestnetInfraAPI interface {
+	// GenerateConfig generates any necessary configuration for the
+	// infrastructure provider during testnet setup.
+	GenerateConfig() error
+
+	// Stop attempts to stop the entire testnet.
+	Stop(ctx context.Context) error
```

Comment:
Could we be a bit more specific about what it means to stop the entire testnet? In docker, these commands will stop the running instances but leave all of the data in place. This doesn't really generalize to digital ocean, where bringing the node 'down' will also remove all of its data.

Reply:
What do you think it should do? From `docker-compose help down`:

```
Stops containers and removes containers, networks, volumes, and images
created by `up`.

By default, the only things removed are:

- Containers for services defined in the Compose file
- Networks defined in the `networks` section of the Compose file
- The default network, if one is used

Networks and volumes defined as `external` are never removed.
```

I'm not sure there can be an equivalent of an "external" volume for DO servers.

The other tricky bit is that the perturbations have this notion of "stopping" a node that purely involves sending it a `SIGTERM`, without removing any resources.

My initial thought was that we should only ever totally destroy infrastructure (implying wiping nodes' data) during `Cleanup`. What do you think? The implication there would be that, for complex tests that rotate new groups of peers in and kill of old groups of peers, we'd end up keeping all VMs running when we may only need some of them.